rustls/src/msgs/deframer/handshake.rs

```diff
@@ -0,0 +1,543 @@
+    ///                 copy_within(from = (17, 23),
+    ///                             to = (12, 18))
+    /// ```
+    ///
+    /// Leaving the buffer and spans:
+    ///
+    /// ```text
+    /// 0 1 2 3 4 5 6 7 8 9 a b c d e f 0 1 2 3 4 5 6 7 8 9
+    /// x x x x x h h h h H H H H H H H H H x x x x x x x x
+    ///           '------------------------'
+    ///            |
+    /// spans = [ { bounds = (5, 18), size = Some(9), .. } ]
+    /// ```
+    pub(crate) fn coalesce(&mut self, containing_buffer: &mut [u8]) {
+        // Strategy: while there is work to do, scan `spans`
+        // for a pair where the first is not complete.  move
+        // the second down towards the first, then reparse the contents.
+        while let Some(i) = self.requires_coalesce() {
+            self.coalesce_one(i, Coalescer::new(containing_buffer));
+        }
+    }
+
+    /// Within `containing_buffer`, move `span[index+1]` to be contiguous
+    /// with `span[index]`.
+    fn coalesce_one(&mut self, index: usize, mut containing_buffer: Coalescer<'_>) {
+        let second = self.spans.remove(index + 1);
+        let mut first = self.spans.remove(index);
+
+        // move the entirety of `second` to be contiguous with `first`
+        let len = second.bounds.len();
+        let target = Range {
+            start: first.bounds.end,
+            end: first.bounds.end + len,
+        };
+
+        containing_buffer.copy_within(second.bounds, target);
+        let delocator = containing_buffer.delocator();
+
+        // now adjust `first` to cover both
+        first.bounds.end += len;
+
+        // finally, attempt to re-dissect `first`
+        let msg = InboundPlainMessage {
+            typ: ContentType::Handshake,
+            version: first.version,
+            payload: delocator.slice_from_range(&first.bounds),
+        };
+
+        for (i, span) in
+            DissectHandshakeIter::new(msg, &delocator.locator(), second.discard).enumerate()
+        {
+            self.spans.insert(index + i, span);
+        }
+    }
+
+    /// We require coalescing if any span except the last is not complete.
+    ///
+    /// Returns an index into `spans` for the first non-complete span:
+    /// this will never be the last item.
+    fn requires_coalesce(&self) -> Option<usize> {
+        self.spans
+            .split_last()
+            .and_then(|(_last, elements)| {
+                elements
+                    .iter()
+                    .enumerate()
+                    .find_map(|(i, span)| (!span.is_complete()).then_some(i))
+            })
+    }
+}
+
+impl Default for HandshakeDeframer {
+    fn default() -> Self {
+        Self {
+            // capacity: a typical upper limit on handshake messages in
+            // a single flight
+            spans: Vec::with_capacity(16),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum HandshakeDeframerStatus {
+    /// We have the given number of messages ready.
+    Ready(usize),
```

Comment:
Thanks for the comments!

Yes, this is unused. I think actually `status()` itself is not really paying its rent, especially since there are other status indicators not included in it. Have eliminated that.